[docs] Entra ID integration (#56930)
gravitational/teleport: docs/pages/identity-governance/entra-id.mdx

---
title: Entra ID Integration
description: Describes how to set up the Teleport Entra ID integration in Teleport.
labels:
 - how-to
 - identity-governance
---

Entra ID integration enables the following features in Teleport:

1. **Single sign-on (SSO):** Configures Teleport authentication with Entra ID as an
   identity provider.
2. **User sync:** Imports Entra ID users as Teleport users.
3. **Group sync:** Imports Entra ID groups as Teleport
   Access Lists. 

> 4. **Application sync (Optional):** Imports Entra ID applications. 

Reviewer: We do not import Applications into teleport. We only import them to access graph.
As opposed to Okta, we only import groups & users into Teleport

> 5. **Integration with Teleport Identity Security (Optional):** Lets you analyze user access paths

Reviewer: Refer that we also import any enterprise app to be displayed there

and policies from the Teleport Identity Security product. 

## How it works

An enterprise application will be configured in the Entra ID. 
The purpose of this application is twofold - 
1. Set up a SAML SSO provider for Teleport. 

> 2. Set up Microsoft graph API credential for Teleport that allows to fetch resources 

Reviewer: This is not accurate if using system credentials and we should enphasize that here to avoid confusion. 

The second case is only true if using OIDC for authentication, otherwise that's not required at all.

from Entra ID directory to Teleport.  

Once the integration is configured, Teleport periodically imports users, groups, 

> and applications into Teleport.

Reviewer: applications are not imported

The installation process provides an option to extend Entra ID integration with the 
Teleport Identity Security product, which offers user [access path and policy analysis](../identity-security/integrations/entra-id.mdx).

### Choosing an authentication method

In order to fetch resources from the Entra ID directory, Teleport needs to be 
configured with credential that grants access to the Microsoft Graph API. 

#### Teleport as an OIDC Provider for Entra ID 

In this setup, Teleport is configured as an OpenID Connect (OIDC) identity provider, 
establishing a trusted connection with an Entra ID application created during setup. 

> This trust allows Teleport to authenticate using the Entra ID application, accessing 

Reviewer: ```suggestion
This trust allows Teleport to generate short-lived credentials to authenticate using the Entra ID application, accessing 
```

permissions tied to it without requiring additional credentials or managed Identities.

Direct bidirectional connectivity between Teleport and Azure is necessary for Azure to 
validate the OIDC tokens issued by Teleport.

#### System credentials

Designed for air-gapped Teleport clusters that are not publicly accessible, this setup 
accommodates environments where Azure cannot validate OIDC tokens issued by Teleport. 

> Instead, Teleport relies on Azure credentials available on the VMs where Teleport 

Reviewer: I think we should rephrase this to avoid referring VMs, first because it works with kubernetes pods and second because we can make those credentials available in AWS/GCP machines

Auth Service is running.

## Prerequisites

- Teleport Identity Governance enabled for your account. Optionally, Identity Security should
also be enabled if you choose to enable Identity Security integration.

> - Your user must have privileged administrator permissions in the Azure account.

Reviewer: ```suggestion
- Your user must have privileged administrator permissions in the Azure EntraID account.
```

This requires Admin for EntraID, not azure

In the integration UI, select the “Microsoft Entra ID” tile. 
![Entra ID Integration](../../img/igs/entraid/entra-tile.png)

In the Microsoft Entra ID configuration UI, you will notice a default integration name “entra-id”
is already populated for you. You will need to select Teleport user(s) that will be assigned as 
the default owner of Access Lists that are created for your Entra ID groups. 

![Entra ID Integration](../../img/igs/entraid/entra-default-owners.png)

In the next step, you will be provided with a Entra ID configuration script.

## Step 2/3: Configure Entra ID

(!docs/pages/includes/identity-governance/azure-shell.mdx!)

## Step 3/3: Finish plugin installation

Once the script is done setting up the necessary permissions,
it prints out tenant ID and client ID. Copy these values and provide it to the
Web UI to finish the integration.

![Entra ID Integration](../../img/igs/entraid/entra-finish.png)

</TabItem>

<TabItem label="Teleport as OIDC provider (tctl)" >

## Step 1/3: Generate configuration script (tctl)

To begin integration, run the `tctl plugins install entraid` command.

```code

> $ tctl plugins install entraid \

Author: I wanted to have them explained in a followup PR where I will be adding the entra ID plugins command in the tctl reference page.  Added a brief descriptions for each flag in https://github.com/gravitational/teleport/pull/56930/commits/bcf67a2e15836f0bb2756c42a45528d9b86a1f77

    --name entra-id-default \
    --auth-connector-name entra-id \
    --default-owner=<Var name="Access List Owner"/> \
    --default-owner=someOtherOwner@teleport.sh  \
    --auth-server <Var name="example.teleport.sh:443" /> 

Step 1: Run the Setup Script
...

Once the script completes, type 'continue' to proceed, 'exit' to quit: 
```

## Step 2/3: Configure Entra ID (tctl)

(!docs/pages/includes/identity-governance/azure-shell.mdx!)

## Step 3/3: Finish plugin installation (tctl)

Once the script is done setting up the necessary permissions,
it prints out tenant ID and client ID. Copy these values and provide it to the
`tctl` to finish the integration.

</TabItem>

<TabItem label="System credential (tctl)" >

## Step 1/4: Set up system credential

You will need to grant Azure Identity with the necessary permissions required for the Entra ID
integration.

> In the Azure Portal, find the identities linked to your Teleport Auth Service VMs,

Reviewer: lets rephrase this with the comments I left before about kube pods and other vms

}

```

</details>

Your identity principal `<Var name="Principal ID" />` now has the necessary permissions to list Applications,
Directories, and Policies.

## Step 2/4: Generate configuration script (tctl)

To begin integration, run the `tctl plugins install entraid` command.

```code
$ tctl plugins install entraid \
    --name entra-id-default \
    --auth-connector-name entra-id \
    --default-owner=<Var name="Access List Owner"/> \
    --default-owner=someOtherOwner@teleport.sh  \
    --auth-server <Var name="example.teleport.sh:443" /> \
    --use-system-credentials 

Step 1: Run the Setup Script
...

Once the script completes, type 'continue' to proceed, 'exit' to quit: 
```

## Step 3/4: Configure Entra ID (tctl)

(!docs/pages/includes/identity-governance/azure-shell.mdx!)

## Step 4/4: Finish plugin installation (tctl)

Once the script is done setting up the necessary permissions,
it prints out tenant ID and client ID. Copy these values and provide it to the
`tctl` to finish the integration.

</TabItem>
</Tabs>

The integration is now configured and the Teleport Entra ID service will start 
importing resources from Entra ID to Teleport.

> ## Manual configuration

Reviewer: This section of the manual is confusing because it's part of the main flow of the documentation, yet it replaces-rather than complements-the other methods described in the "Pages" section. It's also unclear why it starts from step 1/x, which adds to the confusion.

Reviewer 2: Agreed, I don't think we've nailed the UX here yet. This is very hard to follow.

Manual configuration may be preferable if you wish to have more control on the 
Entra ID configuration. Below, an OIDC IdP based manual configuration is 
demonstrated. 

## Step 1/5. Create enterprise application

In the Azure Portal, under “Azure services”, select “Enterprise applications”.
Click on `+ New Application` button, then click `+ Create your own application` button.
Enter a name for your application and create the application.

![Entra ID Integration](../../img/igs/entraid/entra-create-enterprise-app.png)

## Step 2/5. Configure SSO

Open the newly created enterprise application.

Under “Manage” menu select “Single sign-on”. In this configuration UI, you will need to set up Teleport

> as an SAML servicer provider.

Reviewer 2: I'm confused - is this setup using both OIDC and SAML?

Click edit button to configure  “Basic SAML Configuration”.
Enter the SAML assertion endpoint as Entity ID and ACS URL value.
- **Entity ID and ACS URL:** SAML ACS endpoint of your Teleport cluster. 
E.g. `https://<teleport cluster addr>/v1/webapi/saml/acs/entra-id`

For “Attributes & Claims”, attributes with user will already be available for you but you will need to 
add a `groups` claim. 

![Entra ID Integration](../../img/igs/entraid/entra-sso-config.png)

> ## Step 3/5. Configure OIDC IdP

Reviewer: The policy integration also shows the powershell script to create the app and OIDC. This is important if you want to have some kind of reproducible deployment that automatically creates everything in entra id side.

I think you dropped them here but I would rather include them

Author: Its dropped temporarily cause I haven't tested those scripts and wanted to reduce number of changes in this PR. I think its fine to follow up with that later as long as the manual steps in the UI is covered.

Under “App registrations” from Azure services menu, find and open your enterprise
application created in step 1.

Select “Manage > Certificates & secrets” and then select “Federated credentials“.
Click `+ Add credential` button.

Under “Add a credential” UI, configure credential with the following values:
- **Federated credential scenario:** Other issuer

Under “Connect your account”, configure the following values:

> - **Issuer:** https://example.com

Reviewer 2: Someone reading these docs is likely to type the literal "example.com" value in here, which would be wrong. We need to better describe the steps we want the user to take.

- **Type:** Explicit subject identifier
- **Value:** teleport-azure

Under “Credential details”, configure the following values:
- **Name:** teleport-oidc
- **Description:** Teleport OIDC Identity Provider

![Entra ID Integration](../../img/igs/entraid/entra-setup-oidc.png)

## Step 4/5. Configure API permissions

Under the same App registration UI for your enterprise application, select “Manage > API permissions”.

You can add a new graph permission by clicking on `+ Add a permission` button and then selecting 
“Microsoft Graph > Application permissions”.

> The following permissions needs to be added to the application. 

Reviewer 2: ```suggestion
The following permissions need to be added to the application. 
```

- `Application.ReadWrite.OwnedBy`
- `Group.Read.All`
- `User.Read.All`

![Entra ID Integration](../../img/igs/entraid/entra-api-permission.png)

## Step 5/5. Install the Entra ID plugin

Now run the `tctl plugins install entraid` command.

```code
$ tctl plugins install entraid \
    --name entra-id-default \
    --auth-connector-name entra-id \
    --default-owner=<Var name="Access List Owner"/> \
    –-manual-setup
```

> `tctl` will then prompt for Entra ID tenant ID and Application ID of the enterprise 

Reviewer 2: Why is tenant ID lowercase but Application ID is capitalized?

Author: Only because I failed to notice that 😢

application created in step 1.

Once you enter these values, Entra ID plugin will be installed with the OIDC IdP based authentication method.

## FAQ

### What resources are imported to Teleport?

Teleport imports all the users, user groups and its members from the Entra ID directory.
There is no filter available to control custom import rules.

If Teleport Identity Security integration is enabled, Teleport will import 
applications and policies as well. 

### How does it work with nested Access Lists?

If an Entra ID group is assigned as a member to another group, Teleport preserves this assignment 
as a nested Access List. 

However, note that Teleport does not support recursive groups where group A is a member of group B
and group B is also a member of group A.

### What permissions does Teleport need to authenticate with the Microsoft Graph API?

At minimum, Teleport needs read access to users, groups and the main enterprise application 
for which the integration is configured.
```code 
- Application.ReadWrite.OwnedBy
- Group.Read.All
- User.Read.All
```

> If you enable Identity Security integration, you will need a broader sets of permission.

Reviewer 2: ```suggestion
If you enable the Identity Security integration, you will need a broader set of permissions.
```

Do these permissions _replace_ the ones above, or are they needed in addition to the ones above?

Author: Replace it. Added comment to hint that.

#### Teleport as an OIDC Provider for Entra ID 

In this setup, Teleport is configured as an OpenID Connect (OIDC) identity provider, 
establishing a trusted connection with an Entra ID application created during setup. 
This trust allows Teleport to authenticate using the Entra ID application, accessing 
permissions tied to it without requiring additional credentials or managed Identities.

Direct bidirectional connectivity between Teleport and Azure is necessary for Azure to 
validate the OIDC tokens issued by Teleport.

#### System credentials

Designed for air-gapped Teleport clusters that are not publicly accessible, this setup 
accommodates environments where Azure cannot validate OIDC tokens issued by Teleport. 
Instead, Teleport relies on Azure credentials available on the VMs where Teleport 
Auth Service is running.

## Prerequisites

- Teleport Identity Governance enabled for your account. Optionally, Identity Security should
also be enabled if you choose to enable Identity Security integration.
- Your user must have privileged administrator permissions in the Azure account.
- For the OIDC IdP setup, the Teleport cluster must be publicly accessible from the internet.

> - For air gapped clusters, `tctl` must be v16.4.7 or later.

Reviewer 3: Nit: I'd consider removing this line. This guide will only apply to v18+ so a user would likely have a newer version anyway and v16 will be obsolete soon.
```suggestion
```
